WalletWasabi.Tests/UnitTests/MempoolMirrorTests/MirrorMempoolTests.cs

```diff
@@ -0,0 +1,185 @@
+
+				await services.StartAllAsync();
+				await mempoolInstance.TriggerAndWaitRoundAsync(TimeSpan.FromSeconds(7));
+				var localMempoolHashes = mempoolInstance.GetMempoolHashes();
+
+				Assert.Equal(2, localMempoolHashes.Count);
+				Assert.Contains(txid, localMempoolHashes);
+				Assert.Contains(txid2, localMempoolHashes);
+			}
+			finally
+			{
+				await services.StopAllAsync();
+				await coreNode.TryStopAsync();
+			}
+		}
+
+		[Fact]
+		public async Task CanHandleArrivingTxAsync()
+		{
+			var coreNode = await TestNodeBuilder.CreateAsync();
+			using HostedServices services = new();
+			services.Register<MempoolMirror>(new MempoolMirror(TimeSpan.FromSeconds(7), coreNode.RpcClient, coreNode.P2pNode), "Mempool Mirror");
+			try
+			{
+				var rpc = coreNode.RpcClient;
+				var network = rpc.Network;
+				var walletName = "RandomWalletName";
+				await rpc.CreateWalletAsync(walletName);
+
+				await services.StartAllAsync();
+
+				var spendAmount = new Money(0.0004m, MoneyUnit.BTC);
+				await rpc.GenerateAsync(101);
+
+				var txid = await rpc.SendToAddressAsync(BitcoinFactory.CreateBitcoinAddress(network), spendAmount);
+
+				Thread.Sleep(3000);
+
+				var mempoolInstance = services.Get<MempoolMirror>();
+				await mempoolInstance.TriggerAndWaitRoundAsync(TimeSpan.FromSeconds(21));
+				var localMempoolHashes = mempoolInstance.GetMempoolHashes();
+
+				Assert.Single(localMempoolHashes);
+				Assert.Contains(txid, localMempoolHashes);
+			}
+			finally
+			{
+				await services.StopAllAsync();
+				await coreNode.TryStopAsync();
+			}
+		}
+
+		[Fact]
+		public async Task CanHandleTheSameTxSentManyTimesAsync()
+		{
+			var coreNode = await TestNodeBuilder.CreateAsync();
+			using HostedServices services = new();
+			services.Register<MempoolMirror>(new MempoolMirror(TimeSpan.FromSeconds(2), coreNode.RpcClient, coreNode.P2pNode), "Mempool Mirror");
+			try
+			{
+				var rpc = coreNode.RpcClient;
+				var network = rpc.Network;
+				await services.StartAllAsync();
+				var mempoolInstance = services.Get<MempoolMirror>();
+
+				var walletName = "RandomWalletName";
+				await rpc.CreateWalletAsync(walletName);
+
+				using var k1 = new Key();
+				var blockId = await rpc.GenerateToAddressAsync(1, k1.PubKey.WitHash.GetAddress(network));
+				var block = await rpc.GetBlockAsync(blockId[0]);
+				var coinBaseTx = block.Transactions[0];
+
+				var tx = Transaction.Create(network);
+				tx.Inputs.Add(coinBaseTx, 0);
+				tx.Outputs.Add(Money.Coins(49.9999m), BitcoinFactory.CreateBitcoinAddress(network));
+				tx.Sign(k1.GetBitcoinSecret(network), coinBaseTx.Outputs.AsCoins().First());
+				var valid = tx.Check();
+
+				await rpc.GenerateAsync(101);
+
+				for (int i = 0; i < 5; i++)
+				{
+					await rpc.SendRawTransactionAsync(tx);
+
+					await mempoolInstance.TriggerAndWaitRoundAsync(TimeSpan.FromSeconds(7));
+				}
+
+				var localMempoolHashes = mempoolInstance.GetMempoolHashes();
```

Comment:
IMO it's enough to call `await mempoolInstance.TriggerAndWaitRoundAsync(TimeSpan.FromSeconds(7))` outside the loop, you don't have to trigger it every time.

Reply:
Not sure about this.

Reply:
You might be right there. Fixed